### PR TITLE
Fixed STC Bug

### LINKED ISSUE #321: [BUG] ShapeletTransformClassifier doesn't work with y as pd.Series
**Describe the bug**
 ShapeletTransformClassifier doesn't work with y as pd.Series

**To Reproduce**
This breaks: 
```python
from sktime.classification.shapelet_based import ShapeletTransformClassifier
from sktime.datasets import load_arrow_head
from sklearn.model_selection import train_test_split

X, y = load_arrow_head(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y)

c = ShapeletTransformClassifier()
c.fit(X_train, y_train)
```

**Expected behavior**
Works with pd.Series or np.array

**Additional context**
This works: 

```python
from sktime.classification.shapelet_based import ShapeletTransformClassifier
from sktime.datasets import load_arrow_head
from sklearn.model_selection import train_test_split

X, y = load_arrow_head(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y)

c = ShapeletTransformClassifier()
c.fit(X_train, y_train.to_numpy())
```

### PR DESCRIPTION
**Pull request by ABostrom**
#### Reference Issues/PRs
Fixes #321

#### What does this implement/fix? Explain your changes.
Checks if pd.series passed and converts to numpy for internal use.

#### Any other comments?
Added some additional tests for STC although may be unneccessary

### CODE REVIEW

**Comment by mloning** [comment]
Since you're on it, if you have time, it'd be great to make shapelet comply with our interface more generally see #257 :)

**Comment by ABostrom** [comment]
Okay. I'll figure out whatever that error is. I suspect my tests are playing up. I'll have a look at #257 but I'll probably do it tomorrow :)

**Comment by ABostrom** [comment]
@mloning I will get round to this, suddenly got loads of work to juggle at the moment. Hopefully try to get a PR done for end of the week

**Comment by ABostrom** [comment]
Okay @mloning noticed another bug with random seeding which im also fixing to make sure STC is correct. Taking longer than expected.

**Comment by ABostrom** [comment]
@mloning I also noticed in base.py for the Base Classifier the predict method doesn't called check_is_fitted(). See my newest commit

**Comment by mloning** [comment]
@ABostrom if you're feeling brave you can remove all Shapelet related estimators from [this list here](https://github.com/alan-turing-institute/sktime/blob/1e472eaea8e42330e2a4581559b9c981ebf74c87/sktime/tests/_config.py#L50), this will run all standard unit test on them. 

You may have to change the test hyper-params to make it faster (e.g. contract time to 10 secs or so) in the same file in [this dictionary](https://github.com/alan-turing-institute/sktime/blob/1e472eaea8e42330e2a4581559b9c981ebf74c87/sktime/tests/_config.py#L82).

**Comment by ABostrom** [comment]
oooh. scary, will see what happens.

**Comment by ABostrom** [comment]
So I *think* i've resolved the testing issues across all the shapelet family of algorithms. Learned alot about pytest so thanks for that @mloning :)

File: sktime/classification/shapelet_based/_stc.py [hunk 1]
```diff
@@ -84,11 +79,26 @@ def fit(self, X, y):
         -------
         self : object
          """
-        X, y = check_X_y(X, y, enforce_univariate=True)
-        self.n_classes = np.unique(y).shape[0]
-        self.classes_ = class_distribution(np.asarray(y).reshape(-1, 1))[0][0]
+        _X, _y = check_X_y(X, y, enforce_univariate=True)
```
Comment (mloning) [inline sktime/classification/shapelet_based/_stc.py:82]: Why do you create the underscore objects? Should work without the underscores, no?

File: sktime/classification/shapelet_based/_stc.py [hunk 2]
```diff
@@ -84,11 +79,26 @@ def fit(self, X, y):
...
 
-        self.classifier.fit(X, y)
+        # if y is a pd.series then convert to array.
+        if isinstance(_y, pd.Series):
+            _y = _y.to_numpy()
+
+# generate pipeline in fit so that random state can be propogated properly.
+        self.classifier = Pipeline([
```
Comment (mloning) [inline sktime/classification/shapelet_based/_stc.py:89]: I'd call it `self.classifier_`, by sklearn convention, attributes changed in fit have a trailing underscore

File: sktime/tests/_config.py [hunk 3]
```diff
@@ -48,15 +48,15 @@
 
 # TODO fix estimators to pass all tests
 EXCLUDED = [
-    'ContractedShapeletTransform',
+    # 'ContractedShapeletTransform',
```
Comment (mloning) [inline sktime/tests/_config.py:51]: please remove all of the commented lines including the `# 'MrSEQLClassifier'` line to clean this up a bit

File: sktime/classification/shapelet_based/_stc.py (hunk 1 above)
Comment (ABostrom) [reply sktime/classification/shapelet_based/_stc.py:82]: created the underscores as that was a suggestion you left me before about not overwriting input args
I'm happy to remove that.

Comment (ABostrom) [reply sktime/classification/shapelet_based/_stc.py:82]: actually there was also a weird error, i think with the fact X was getting shuffled so i was getting an idemptoency failure because the shapes were mismatched.

**Comment by ABostrom** [comment]
@mloning @jasonlines So I think the current build errors, show case the current differences between my tsml implementation and the sktime one. Between the same run of the same datasets you cannot guarantee that you will evaluate the same amount of shapelets in the same time frame. Especially on different architectures etc. Which results in idemptoency failures and inconsistent outputs on the same seeds / datasets. 

Might need to re-include them in the exclusion list, and review this particular check for future contract approaches.

**Comment by mloning** [comment]
@ABostrom does this only apply to the contracted version? Or also the `ShapeletTransformClassifier`?

**Comment by ABostrom** [comment]
@mloning shapelet transform classifier uses the contracted shapelets in a pipeline with radnom forest

File: sktime/classification/shapelet_based/_stc.py (hunk 1 above)
Comment (mloning) [reply sktime/classification/shapelet_based/_stc.py]: ah sorry I meant not overwriting hyper-parameters set in the constructor, input args can be overwritten

**Comment by mloning** [comment]
Let's wait if @jasonlines has any idea how to fix this, otherwise we could try to write some code to skip/ignore the idempotency test for shapelets

**Comment by ABostrom** [comment]
@mloning I agree, I will have a little look at making some changes to _config.py and test_all_estimators to have an exclusion list for certain estimators.

**Comment by mloning** [comment]
@ABostrom decorators may work and may be the cleanest solution, something along the following lines:
```python
@skip_if(estimator=[ShapeletTransform])
def check_idempotency():
    ....
```
Pytest may even have decorators for that already

**Comment by ABostrom** [comment]
my preference and ofc feel free to overrule on this is that all config for tests should stay in config. Seems like with decorators, whilst addmittedly a clean solution could make it difficult where to track down or reconfigure tests in the future. 

`
EXCLUDED_FROM_TESTS = {
    "ShapeletTransformClassifier": [check_fit_idempotent],
    "ContractedShapeletTransform": [check_fit_idempotent],
}
`

**Comment by mloning** [comment]
Sounds good to me!

**Comment by ABostrom** [comment]
@mloning I'm not sure how this works. If i change the config / pytests would that run my version of the tests or the old versions through travis etc?

**Comment by mloning** [comment]
It'll run whatever is on the PR I think, of course you can run them locally first to see if everything is working, so feel free to change things and push to the same branch. You also need to sync with the latest changes on dev.

**Comment by ABostrom** [comment]
Yeah they have all been passing locally. Might be I'm being stupid. will debug it.

**Comment by ABostrom** [comment]
changed the config params for ST to make sure it can actually finish in a time manner suitable for the build/CI server

**Comment by ABostrom** [comment]
@mloning if you can check it over Markus. I think it's looking pretty good. Turned into a bit more than some "simple" shapelet fix. As always these things do :)

File: sktime/classification/shapelet_based/_stc.py [hunk 4]
```diff
@@ -84,19 +79,27 @@ def fit(self, X, y):
...
-        #        self.st_X =self.shapelet_transform.transform(X)
-        #        print("Transform complete")
-        #        X = np.asarray([a.values for a in X.iloc[:, 0]])
-        #        self.classifier.fit(X,y)
-        #       print("Build classifier complete")
+        X, _y = check_X_y(X, y, enforce_univariate=True)
+
+        # if y is a pd.series then convert to array.
+        if isinstance(_y, pd.Series):
+            _y = _y.to_numpy()
```
Comment (mloning) [inline sktime/classification/shapelet_based/_stc.py:86]: why not also simply have `y` instead of `_y`?

File: sktime/classification/shapelet_based/tests/test_stc.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+
+# import numpy as np
+# from sklearn.model_selection import train_test_split
+# from sktime.classification.shapelet_based import ShapeletTransformClassifier
+# from sktime.datasets import load_italy_power_demand
+
+
+# def test_stc_with_pd():
```
Comment (mloning) [inline sktime/classification/shapelet_based/tests/test_stc.py:8]: please remove the file if no longer needed - if we want to support y as a np.array we should probably add this to our collection of estimator checks and run it on all

File: sktime/transformers/series_as_features/shapelets.py [hunk 6]
```diff
@@ -118,38 +118,42 @@ def fit(self, X, y=None):
         self : FullShapeletTransform
             This estimator
         """
-        X = check_X(X, enforce_univariate=True)
+        _X, _y = check_X_y(X, y, enforce_univariate=True)
```
Comment (mloning) [inline sktime/transformers/series_as_features/shapelets.py:121]: same here, having `X` and `y` should be fine no (without the underscores)?

Comment (mloning) [reply sktime/transformers/series_as_features/shapelets.py:121]: just makes it consistent with most other code in sklearn and sktime

File: sktime/utils/_testing/estimator_checks.py [hunk 7]
```diff
@@ -53,7 +53,7 @@
 ]
 
 
-def check_estimator(Estimator):
+def check_estimator(Estimator, Check_Exclusions=None):
```
Comment (mloning) [inline sktime/utils/_testing/estimator_checks.py:56]: for args please use lower case, so `check_exclusions` or simply `exclude`

File: sktime/utils/_testing/estimator_checks.py [hunk 8]
```diff
@@ -66,7 +66,10 @@ def check_estimator(Estimator):
         If Estimator does not comply
     """
     for check in yield_estimator_checks():
-        check(Estimator)
+
+        # check if associated test is not included in the exclusion list
+        if not Check_Exclusions.__contains__(check.__name__):
```
Comment (mloning) [inline sktime/utils/_testing/estimator_checks.py:71]: Why not this? :) 
```suggestion
        if check.__name__ not in Check_Exclusions:
```

File: sktime/tests/_config.py [hunk 9]
```diff
@@ -16,49 +17,44 @@
...
+from sktime.transformers.series_as_features.summarize import (
+    FittedParamExtractor, TSFreshFeatureExtractor,
+    TSFreshRelevantFeatureExtractor)
 from sktime.transformers.single_series.adapt import \
     SingleSeriesTransformAdaptor
 from sktime.transformers.single_series.detrend import Detrender
-from sktime.transformers.series_as_features.interpolate import TSInterpolator
 
 # TODO fix estimators to pass all tests
 EXCLUDED = [
```
Comment (mloning) [inline sktime/tests/_config.py:45]: maybe rename to `EXCLUDED_ESTIMATORS`

File: sktime/tests/_config.py [hunk 10]
```diff
@@ -16,49 +17,44 @@
...
     'KNeighborsTimeSeriesClassifier',
-    # 'MrSEQLClassifier',
     'ProximityForest',
     'ProximityStump',
     'ProximityTree',
-    'ShapeletTransform',
-    'ShapeletTransformClassifier',
 ]
 
+EXCLUDED_FROM_TESTS = {
```
Comment (mloning) [inline sktime/tests/_config.py:53]: and this one to `EXCLUDED_TESTS`

**Comment by ABostrom** [comment]
Will do these now. Was just sorting a PR for sktime-dl. but now thats done, I'll tidy this up.

**Comment by mloning** [comment]
@ABostrom looks all good to me now! Thanks for the clean up of shapelets! :) Ready to merge when you are.

**Comment by ABostrom** [comment]
good to go Markus. Think it looks pretty comprehensive.
